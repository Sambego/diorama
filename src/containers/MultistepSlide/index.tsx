import DeckContext, { DeckContextInternal } from "@contexts/DeckContext";
import MultiSlideContext, {
	MultiSlideInnerContext,
} from "@contexts/MultiSlideContext";
import React, {
	useCallback,
	useContext,
	useEffect,
	useMemo,
	useState,
} from "react";
import Slide from "../Slide";

export default function MultistepSlide({
	render,
	stepsCount,
	...props
}: React.ComponentProps<typeof Slide> & {
	render?: (step: number, totalSteps: number) => React.ReactNode;
	stepsCount: (() => number) | number;
}) {
	const multislideCxt = useContext(MultiSlideInnerContext);
	const ctx = useContext(DeckContext);
	const intDeckCtx = useContext(DeckContextInternal);

	const [currentStep, setCurrentStep] = useState(0);
	const totalSteps = useMemo(() => {
		if (typeof stepsCount === "function") {
			return stepsCount();
		}
		return stepsCount;
	}, [stepsCount]);

	useEffect(() => {
		setCurrentStep(0);
	}, [ctx.currentSlideIndex]);

	const listener = useCallback(
		(direction: "next" | "prev") => {
			if (direction === "prev") {
				if (currentStep > 0) {
					setCurrentStep(c => c - 1);
					return true;
				}
			} else if (currentStep < totalSteps - 1) {
				setCurrentStep(c => c + 1);
				return true;
			}
			return false;
		},
		[totalSteps, currentStep, setCurrentStep]
	);

	useEffect(() => {
		multislideCxt.setMultistepSlideHandler(direction => listener(direction));
		return () => {
			multislideCxt.setMultistepSlideHandler(null);
		};
	}, [listener]);

	const innerCtxValue = useMemo(
		() => ({
			currentSlide: currentStep,
			totalSlides: totalSteps,
			isMultiSlide: true,
		}),
		[totalSteps, currentStep]
	);

	useEffect(() => {
		intDeckCtx.setMultiSlideInfo(innerCtxValue);
		return () => {
			intDeckCtx.setMultiSlideInfo(null);
		};
	}, [innerCtxValue]);

	return (
		// eslint-disable-next-line react/jsx-props-no-spreading
		<Slide {...props}>
			<MultiSlideContext.Provider value={innerCtxValue}>
				{render && render(currentStep, totalSteps)}
				{props.children}
			</MultiSlideContext.Provider>
		</Slide>
	);
}
