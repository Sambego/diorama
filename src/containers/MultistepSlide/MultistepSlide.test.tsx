import React from "react";
import renderer from "react-test-renderer";
import MultistepSlide from ".";

it("renders correctly using render prop", () => {
	const rtr = renderer.create(
		<MultistepSlide
			stepsCount={2}
			render={(step, totalSteps) => {
				return <div>Show step {step + 1}</div>;
			}}
		/>
	);
	const tree = rtr.toJSON();
	expect(tree).toMatchSnapshot();
});
