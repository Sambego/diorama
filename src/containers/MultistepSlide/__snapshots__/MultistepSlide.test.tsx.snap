// Vitest Snapshot v1

exports[`renders correctly using render prop 1`] = `
<div
  className="_slide_1aodf_1 diorama-slide "
  style={{}}
>
  <div
    className="diorama-content _content_1aodf_6"
  >
    <div>
      Show step 
      1
    </div>
  </div>
</div>
`;
